.NET/Microsoft.Recognizers.Text/Recognizer.cs

```diff
@@ -1,47 +1,39 @@
-﻿using System;
-using System.Collections.Generic;
-
-namespace Microsoft.Recognizers.Text
-{
-    public abstract class Recognizer : IRecognizer {
-
-        private const string NoneOptions = "None";
-
-        private readonly ModelContainer modelContainer = new ModelContainer();
-
-        protected void RegisterModel(string culture, Type type, string options, IModel model)
-        {
-            modelContainer.RegisterModel(culture, type, model, options);
-        }
-
-        protected void RegisterModel(string culture, string options, Dictionary<Type, IModel> models)
-        {
-            modelContainer.RegisterModel(culture, models, options);
-        }
-
-        public IModel GetModel<TModel>(string culture, bool fallbackToDefaultCulture = true, string options = NoneOptions)
-        {
-            return modelContainer.GetModel<TModel>(culture, fallbackToDefaultCulture, options);
-        }
-
-        public bool TryGetModel<TModel>(string culture, out IModel model, bool fallbackToDefaultCulture = true, string options = NoneOptions)
-        {
-            return modelContainer.TryGetModel<TModel>(culture, out model, fallbackToDefaultCulture, options);
-        }
-
-        protected bool ContainsModels()
-        {
-            return modelContainer.ContainsModels();
-        }
-
-        public bool ContainsModel<TModel>(string culture, bool fallbackToDefaultCulture = true, string options = NoneOptions)
-        {
-            return modelContainer.ContainsModel<TModel>(culture, fallbackToDefaultCulture, options);
-        }
-
-        protected IModel GetSingleModel<TModel>()
-        {
-            return modelContainer.GetSingleModel<TModel>();
-        }
-    }
-}
+﻿using System;
+
+namespace Microsoft.Recognizers.Text
+{
+    public abstract class Recognizer<TModelOptions>
+    {
+        private readonly ModelFactory<TModelOptions> factory;
+
+        public string CurrentCulture { get; private set; }
```

Comment:
This new refactor doesn't cover all previous requirements. 
One Recognizer instance can contain models for multiple cultures, so CurrentCulture should belong in each model, not per-recognizer.
Moreover one of the key previously supported functionalities is now missing. If a recognizer class is initialized with no culture parameters, models for all supported cultures should be created and kept in its internal cache. This is a behaviour utilized by many consumers of the project that we can't remove.

Reply:
What about instead of exposing the recognizer constructors, have clear factory methods? 
Something like var recog = CreateRecognizer(culture) and CreateRecognizer(fallbackCulture). The first returns a single specific culture and the second an instance with all supported cultures and possibly a fallback culture?
Names are only to illustrate the idea. Parameters could be passed differently. And method names + code documentation would make each behaviour explicit.

Reply:
We updated the recognizer constructors to support the `fallbackCulture` scenario and to contain multiple culture models.

```C#
string DefaultCulture = Culture.English;
var recognizer = new DateTimeRecognizer(DefaultCulture);
Assert.AreEqual(recognizer.GetDateTimeModel(DefaultCulture), recognizer.GetDateTimeModel());
Assert.AreNotEqual(recognizer.GetDateTimeModel(Culture.Spanish), recognizer.GetDateTimeModel());
```

Also, by default the recognizer uses lazy model initialization which can be turned off using the `lazyInitialization` parameter, forcing the recognizer to initialize all registered models:

```C#
var recognizer = new DateTimeRecognizer(DefaultCulture, DateTimeOptions.None, lazyInitialization: false);
```